File: backend/src/gravity_ai/core/contracts.py
from __future__ import annotations

from dataclasses import asdict, dataclass, field, is_dataclass
from datetime import datetime, timezone
from enum import Enum, StrEnum
from typing import Any, Mapping
from uuid import uuid4


class RiskLevel(StrEnum):
    SAFE = "safe"
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    DESTRUCTIVE = "destructive"


class ToolPermission(StrEnum):
    FILESYSTEM_READ = "filesystem.read"
    FILESYSTEM_WRITE = "filesystem.write"
    FILESYSTEM_DELETE = "filesystem.delete"
    SYSTEM_PROCESS = "system.process"
    SYSTEM_SETTINGS = "system.settings"
    NETWORK = "network"
    BROWSER = "browser"
    AUTOMATION = "automation"
    MEMORY_READ = "memory.read"
    MEMORY_WRITE = "memory.write"


class ToolStatus(StrEnum):
    SUCCESS = "success"
    ERROR = "error"
    REQUIRES_CONFIRMATION = "requires_confirmation"
    SKIPPED = "skipped"


@dataclass(frozen=True)
class ToolDefinition:
    name: str
    description: str
    parameters_schema: Mapping[str, Any] = field(default_factory=dict)
    permissions: tuple[ToolPermission, ...] = field(default_factory=tuple)
    risk: RiskLevel = RiskLevel.LOW
    requires_confirmation: bool = False

    def to_dict(self) -> dict[str, Any]:
        return _jsonable(asdict(self))


@dataclass(frozen=True)
class ToolCall:
    tool_name: str
    arguments: Mapping[str, Any] = field(default_factory=dict)
    confirmed: bool = False
    call_id: str = field(default_factory=lambda: str(uuid4()))
    requested_at: str = field(default_factory=lambda: utc_now())

    def to_dict(self) -> dict[str, Any]:
        return _jsonable(asdict(self))


@dataclass(frozen=True)
class ToolResult:
    tool_name: str
    status: ToolStatus
    content: Mapping[str, Any] = field(default_factory=dict)
    error: str | None = None
    call_id: str | None = None
    completed_at: str = field(default_factory=lambda: utc_now())

    @classmethod
    def ok(
        cls,
        tool_name: str,
        content: Mapping[str, Any] | None = None,
        call_id: str | None = None,
    ) -> "ToolResult":
        return cls(
            tool_name=tool_name,
            status=ToolStatus.SUCCESS,
            content=content or {},
            call_id=call_id,
        )

    @classmethod
    def error_result(
        cls,
        tool_name: str,
        error: str,
        call_id: str | None = None,
    ) -> "ToolResult":
        return cls(
            tool_name=tool_name,
            status=ToolStatus.ERROR,
            error=error,
            call_id=call_id,
        )

    @classmethod
    def confirmation_required(
        cls,
        tool_name: str,
        reason: str,
        call_id: str | None = None,
    ) -> "ToolResult":
        return cls(
            tool_name=tool_name,
            status=ToolStatus.REQUIRES_CONFIRMATION,
            error=reason,
            call_id=call_id,
        )

    def to_dict(self) -> dict[str, Any]:
        return _jsonable(asdict(self))


def utc_now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _jsonable(value: Any) -> Any:
    if isinstance(value, Enum):
        return value.value
    if is_dataclass(value):
        return _jsonable(asdict(value))
    if isinstance(value, dict):
        return {str(key): _jsonable(item) for key, item in value.items()}
    if isinstance(value, (list, tuple, set)):
        return [_jsonable(item) for item in value]
    return value

File: packages/shared/src/contracts.ts
export const RiskLevel = {
  Safe: "safe",
  Low: "low",
  Medium: "medium",
  High: "high",
  Destructive: "destructive",
} as const;

export type RiskLevel = (typeof RiskLevel)[keyof typeof RiskLevel];

export const ToolStatus = {
  Success: "success",
  Error: "error",
  RequiresConfirmation: "requires_confirmation",
  Skipped: "skipped",
} as const;

export type ToolStatus = (typeof ToolStatus)[keyof typeof ToolStatus];

export const ToolPermission = {
  FilesystemRead: "filesystem.read",
  FilesystemWrite: "filesystem.write",
  FilesystemDelete: "filesystem.delete",
  SystemProcess: "system.process",
  SystemSettings: "system.settings",
  Network: "network",
  Browser: "browser",
  Automation: "automation",
  MemoryRead: "memory.read",
  MemoryWrite: "memory.write",
} as const;

export type ToolPermission = (typeof ToolPermission)[keyof typeof ToolPermission];

export interface ToolDefinition {
  name: string;
  description: string;
  parameters_schema: Record<string, unknown>;
  permissions: ToolPermission[];
  risk: RiskLevel;
  requires_confirmation: boolean;
}

export interface ToolCall {
  toolName: string;
  arguments: Record<string, unknown>;
  confirmed: boolean;
}

export interface ToolResult {
  tool_name: string;
  status: ToolStatus;
  content: Record<string, unknown>;
  error: string | null;
  call_id: string | null;
  completed_at: string;
}

export type PluginPermission = ToolPermission | string;

export interface PluginCommand {
  name: string;
  description: string;
  tool: string;
}

export interface PluginManifest {
  id: string;
  name: string;
  version: string;
  entrypoint: string;
  permissions: PluginPermission[];
  commands: PluginCommand[];
  settings: Record<string, unknown>;
}

export interface ChatMessage {
  role: "system" | "user" | "assistant" | "tool";
  content: string;
  metadata?: Record<string, string>;
}

export interface ModelResponse {
  content: string;
  model: string;
  provider: string;
  usage: Record<string, number>;
}

export const MemoryScope = {
  ShortTerm: "short_term",
  LongTerm: "long_term",
  Preference: "preference",
  History: "history",
  Context: "context",
} as const;

export type MemoryScope = (typeof MemoryScope)[keyof typeof MemoryScope];

export interface MemoryEntry {
  entry_id: string;
  scope: MemoryScope;
  key: string;
  value: string;
  metadata: Record<string, string>;
  created_at: string;
  updated_at: string;
}

export interface MemoryQuery {
  scope?: MemoryScope;
  text?: string;
  limit?: number;
}

export interface HealthResponse {
  name: string;
  version: string;
  status: "ok";
  rootDir: string;
  tools: number;
  plugins: number;
}

export interface ChatResponse {
  message: string;
  model: string;
  toolSuggestions: string[];
  memoryMatches: Array<{ scope: string; key: string; value: string }>;
}

